### ISSUE TITLE
Build failure: Failed to initialize profile ('d1x', ''): C++ compiler does not support class template argument deduction.

### ISSUE THREAD

**Issue by Thenuke-Alia-Immortalis** [dxx-rebirth/dxx-rebirth#780]
Hi, I'm new on both Linux and compiling repos by myself so I may have done something wrong while building DXX-Rebirth. I had to read the install guidelines a few times to understand what I needed to do.

Specs : Linux Mint 21.3
x86\_64 (64-bit Intel/AMD)

Step by step, I did this :

-Download the source code by clicking on Download .zip : `https://github.com/dxx-rebirth/dxx-rebirth/archive/refs/heads/master.zip`
-Run this in the terminal to download/update the prerequisites : `apt-get install build-essential scons libsdl1.2-dev libsdl-image1.2-dev libsdl-mixer1.2-dev libphysfs-dev`
-Unzipped the source code in a folder, open it and open a terminal from it.
-Run this : `scons`

I don't think I needed to provide any build arguments except whether I wanted D1X or D2X, either way it fails to build and tells me `Failed to initialize profile ('d1x', ''): C++ compiler does not support class template argument deduction.` I can't read code nor error logs so I don't know what it all means.

I'm afraid I don't know where to find site-local.py.

[sconf.log](https://github.com/user-attachments/files/15760256/sconf.log)

```~/Bureau/dxx-rebirth-master$ scons
scons: Reading SConscript files ...
===== D1X-Rebirth v0.61.0 0.60.0-beta2-3003-g10a62eb6451e archive_10a62eb6451e6c2a7213a7051bb3753c52c98ec8 =====
===== D2X-Rebirth v0.61.0 0.60.0-beta2-3003-g10a62eb6451e archive_10a62eb6451e6c2a7213a7051bb3753c52c98ec8 =====
DXX-Archive.3: compiling on 'linux'/'x86_64' for 'linux' into build/
DXX-Archive.3: checking version of C++ compiler 'g++' ... 'g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0'
DXX-Archive.3: checking path to assembler ... b'as'
DXX-Archive.3: checking version of assembler b'as' ... 'Assembleur GNU (GNU Binutils for Ubuntu) 2.38'
DXX-Archive.3: checking path to linker ... b'ld'
DXX-Archive.3: checking version of linker b'ld' ... 'GNU ld (GNU Binutils for Ubuntu) 2.38'
DXX-Archive.3: CHOST: None
DXX-Archive.3: CXX: 'g++'
DXX-Archive.3: CPPDEFINES: [('PHYSFS_DEPRECATED', ''), ('DXX_USE_SHAREPATH', 1)]
DXX-Archive.3: CPPPATH: ['common/include', 'common/main', '.', 'build/']
DXX-Archive.3: CPPFLAGS: ['-Wno-sign-compare']
DXX-Archive.3: CXXFLAGS: ['-g', '-O2', '-ftabstop=4', '-Wall', '-Werror=extra', '-Werror=format=2', '-Werror=missing-braces', '-Werror=missing-include-dirs', '-Werror=uninitialized', '-Werror=undef', '-Werror=pointer-arith', '-Werror=cast-qual', '-Werror=missing-declarations', '-Werror=vla', '-pthread', '-funsigned-char']
DXX-Archive.3: LIBS: None
DXX-Archive.3: LINKFLAGS: []
DXX-Archive.3: $CCACHE_PREFIX: None
DXX-Archive.3: $DISTCC_HOSTS: None
DXX-Archive.3: checking whether C++ compiler and linker work...yes
DXX-Archive.3: checking whether C++ compiler accepts -std=gnu++20...yes
DXX-Archive.3: checking endian to use...little
DXX-Archive.3: checking whether to enable word alignment fixups...no
DXX-Archive.3: building with OpenGL
DXX-Archive.3: checking whether to enable release options...yes
DXX-Archive.3: checking whether to enable memory allocation tracking...no
DXX-Archive.3: checking whether to enable level editor...no
DXX-Archive.3: checking whether to enable IPv6 support...no
DXX-Archive.3: checking whether to enable stereoscopic rendering...yes
DXX-Archive.3: checking whether to enable multiplayer over UDP...yes
DXX-Archive.3: checking whether to enable UDP game tracker...yes
DXX-Archive.3: checking how to handle ADL MIDI...disabled
DXX-Archive.3: checking how to format screenshots...png
DXX-Archive.3: checking libpng pkg-config libpng
DXX-Archive.3: using pkg-config at discovered path /usr/bin/pkg-config
DXX-Archive.3: reading libpng version from `/usr/bin/pkg-config --modversion libpng`
DXX-Archive.3: libpng version: '1.6.37'
DXX-Archive.3: reading libpng settings from `/usr/bin/pkg-config --cflags --libs libpng`
DXX-Archive.3: libpng settings: {'CPPPATH': ['/usr/include/libpng16'], 'LIBS': ['png16', 'z']}
DXX-Archive.3: checking for usable library png...yes
DXX-Archive.3: checking physfs pkg-config physfs
DXX-Archive.3: reading physfs version from `/usr/bin/pkg-config --modversion physfs`
DXX-Archive.3: physfs version: '3.0.2'
DXX-Archive.3: reading physfs settings from `/usr/bin/pkg-config --cflags --libs physfs`
DXX-Archive.3: physfs settings: {'LIBS': ['physfs']}
DXX-Archive.3: checking for usable library physfs...yes
DXX-Archive.3: checking for usable library ['GL', 'GLU']...yes
DXX-Archive.3: checking SDL pkg-config sdl
DXX-Archive.3: reading sdl version from `/usr/bin/pkg-config --modversion sdl`
DXX-Archive.3: SDL version: '1.2.15'
DXX-Archive.3: reading SDL settings from `/usr/bin/pkg-config --cflags --libs sdl`
DXX-Archive.3: SDL settings: {'CPPDEFINES': [['_GNU_SOURCE', '1'], '_REENTRANT'], 'CPPPATH': ['/usr/include/SDL'], 'LIBS': ['SDL']}
DXX-Archive.3: checking whether to enable joystick support...yes
DXX-Archive.3: checking for usable library SDL with OpenGL...yes
DXX-Archive.3: checking whether to use SDL_image...yes
DXX-Archive.3: checking SDL_image pkg-config SDL_image
DXX-Archive.3: reading SDL_image version from `/usr/bin/pkg-config --modversion SDL_image`
DXX-Archive.3: SDL_image version: '1.2.12'
DXX-Archive.3: reading SDL_image settings from `/usr/bin/pkg-config --cflags --libs SDL_image`
DXX-Archive.3: SDL_image settings: {'CPPDEFINES': [['_GNU_SOURCE', '1'], '_REENTRANT'], 'CPPPATH': ['/usr/include/SDL'], 'LIBS': ['SDL_image', 'SDL']}
DXX-Archive.3: checking for usable library SDL_image...yes
DXX-Archive.3: checking whether to use SDL_mixer...yes
DXX-Archive.3: checking SDL_mixer pkg-config SDL_mixer
DXX-Archive.3: reading SDL_mixer version from `/usr/bin/pkg-config --modversion SDL_mixer`
DXX-Archive.3: SDL_mixer version: '1.2.12'
DXX-Archive.3: reading SDL_mixer settings from `/usr/bin/pkg-config --cflags --libs SDL_mixer`
DXX-Archive.3: SDL_mixer settings: {'CPPDEFINES': [['_GNU_SOURCE', '1'], '_REENTRANT'], 'CPPPATH': ['/usr/include/SDL'], 'LIBS': ['SDL_mixer', 'SDL']}
DXX-Archive.3: checking for usable library SDL_mixer...yes
DXX-Archive.3: checking whether compiler optimizes function __attribute__((__error__))...yes
DXX-Archive.3: checking whether compiler implements __builtin_bswap{16,32,64} functions...yes
DXX-Archive.3: checking whether compiler optimizes __builtin_constant_p...yes
DXX-Archive.3: checking whether compiler accepts __builtin_expect...yes
DXX-Archive.3: checking whether compiler accepts __builtin_FILE, __builtin_LINE...yes
DXX-Archive.3: checking whether compiler optimizes __builtin_object_size...yes
DXX-Archive.3: checking whether compiler understands embedded compound statements...yes
DXX-Archive.3: checking for function __attribute__((always_inline))...yes
DXX-Archive.3: checking for function __attribute__((alloc_size))...yes
DXX-Archive.3: checking for function __attribute__((cold))...yes
DXX-Archive.3: checking for function __attribute__((format_arg))...yes
DXX-Archive.3: checking for function __attribute__((format(printf)))...yes
DXX-Archive.3: checking for function __attribute__((malloc))...yes
DXX-Archive.3: checking for function __attribute__((nonnull))...yes
DXX-Archive.3: checking for function __attribute__((used))...yes
DXX-Archive.3: checking for function __attribute__((unused))...yes
DXX-Archive.3: checking whether compiler optimizes function __attribute__((__warning__))...yes
DXX-Archive.3: checking for C++11 intrinsic static_assert when true...yes
DXX-Archive.3: checking for C++11 intrinsic static_assert when false...yes
DXX-Archive.3: checking whether compiler handles classes from "using namespace"...yes
DXX-Archive.3: checking for required C++11, C++14, C++17, C++20 standard features...no
DXX-Archive.3: checking for C++20 class template argument deduction...no
DXX-Archive.3: checking for C++20 abbreviated function templates...yes
DXX-Archive.3: checking for C++20 explicitly defaulted operator==...yes
DXX-Archive.3: checking for C++20 std::make_unique_for_overwrite...yes
DXX-Archive.3: checking for C++20 requires clause...yes
DXX-Archive.3: checking for C++20 std::ranges...yes
DXX-Archive.3: checking for C++20 std::span...yes
DXX-Archive.3: checking for C++17 constexpr if...yes
DXX-Archive.3: checking for C++17 fold expressions...yes
DXX-Archive.3: checking for C++17 structured binding declarations...yes
DXX-Archive.3: checking for C++17 attribute [[fallthrough]]...yes
DXX-Archive.3: checking for C++17 attribute [[nodiscard]]...yes
DXX-Archive.3: checking for C++14 template variables...yes
DXX-Archive.3: checking for C++14 std::exchange...yes
DXX-Archive.3: checking for C++14 std::index_sequence...yes
DXX-Archive.3: checking for C++14 std::make_unique...yes
DXX-Archive.3: checking for C++11 std::addressof...yes
DXX-Archive.3: checking for C++11 constexpr...yes
DXX-Archive.3: checking for C++11 nullptr...yes
DXX-Archive.3: checking for C++11 explicit operator bool...yes
DXX-Archive.3: checking for C++11 explicitly deleted functions...yes
DXX-Archive.3: checking for C++11 template aliases...yes
DXX-Archive.3: checking for C++11 trailing function return type...yes
DXX-Archive.3: checking for C++11 class scope static constexpr assignment...yes
DXX-Archive.3: checking for C++11 braced base class initialization...yes
DXX-Archive.3: checking for C++11 std::array...yes
DXX-Archive.3: checking for C++11 std::begin(), std::end()...yes
DXX-Archive.3: checking for C++11 range-based for...yes
DXX-Archive.3: checking for C++11 <type_traits>...yes
DXX-Archive.3: checking for C++11 std::unordered_map::emplace...yes
DXX-Archive.3: checking for C++11 reference qualified methods...yes
DXX-Archive.3: checking for C++11 inheriting constructors...yes
scons: *** Failed to initialize profile ('d1x', ''): C++ compiler does not support class template argument deduction.  See build/sconf.log for details.  Stop.```

**Comment by vLKp** [comment]
Your procedure for obtaining the source is fine, though I would have used a `git clone` so that you can pull updated code without needing to redownload the zip file each time.

Yes, you can skip providing arguments to `scons`, if the defaults are acceptable for you.  `site-local.py` is a way to set Rebirth-specific scons options without the need to write them on the command line.  If not present, it is treated as empty.  Since you did not need to set any options, the lack of `site-local.py` is fine.

You made one significant mistake in the environment setup.  Per `Documentation/c++std.markdown`, the minimum required version of gcc is [gcc-12](https://gcc.gnu.org/gcc-12/) (released 2022-05-06).  Your output says that your system compiler is gcc-11.  I know that gcc-11 is too old to support Rebirth, and the specific error you got is one example of where gcc-11 is not acceptable, but gcc-12 would work.  The error you saw is specifically designed to abort the build early, with a (relatively) clear error message, rather than letting the compiler try and fail to parse those features in the code.

---

Please install gcc-12, then try using `scons CXX=g++12`.  That argument is mandatory, since a bare `g++` on your system refers to gcc-11, and installing gcc-12 will (probably) not change that.  If you still have problems, report back, and post the `sconf.log` from the new run.

Based on some searches for `linux mint gcc-12`, I suggest trying `apt-get install g++-12`.  https://forums.linuxmint.com/viewtopic.php?p=2466933&sid=31f992dc07800eb6f26064019552d2d8#p2466933 seems to suggest that gcc-13 (note, 13, not 12) is not practical for you to get, but that gcc-12 should already be readily available to you.

**Comment by Thenuke-Alia-Immortalis** [comment]
Thank you for the detailed answer.

It's still giving me the same error even though I just installed gcc-12 as you said.

`g++-12 --version
g++-12 (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
`
When I try `scons CXX=g++12`, I get this :

```
scons CXX=g++12
scons: Reading SConscript files ...
===== D1X-Rebirth v0.61.0 0.60.0-beta2-3003-g10a62eb6451e archive_10a62eb6451e6c2a7213a7051bb3753c52c98ec8 =====
===== D2X-Rebirth v0.61.0 0.60.0-beta2-3003-g10a62eb6451e archive_10a62eb6451e6c2a7213a7051bb3753c52c98ec8 =====
DXX-Archive.3: compiling on 'linux'/'x86_64' for 'linux' into build/
DXX-Archive.3: checking version of C++ compiler 'g++12' ... error: No such file or directory
scons: *** Failed to initialize profile ('d1x', ''): Failed to run 'g++12'.  See build/sconf.log for details.  Stop.
```

[sconf.log](https://github.com/user-attachments/files/15784612/sconf.log)

**Comment by vLKp** [comment]
I made a mistake in my instructions above.  I wrote `scons CXX=g++12`, which you used exactly.  However, your shell output shows that the compiler is named `g++-12`, not `g++12` (missing dash in my version).  I knew gcc names have the dash in there, but failed to write it in my prior instructions.  Use `scons CXX=g++-12`, since you want the right side to be exactly the name of the compiler you want to run.
